Add tests more test for typing rules that involve wires

thiskappaisgrey opened the pull request.

gussmith23 commented:
You *should* be able to delete the wires before running the typing rules...if that's not the case, there's a bug somewhere! 

The wire typing rule is definitely still needed, bu if you have time, can you help me debug why the wires need to be there for typing to work? That shouldn't be the case!

Also, are you finding that the expression you're testing gets the _wrong_ type? Or is it getting no type at all?

gussmith23 commented:
It looks like in the specific case of the test you uploaded:
1. The wires can be deleted before typechecking
2. It's actually the rule for Xor that's missing

Can you keep the Wire rule, but add the Xor rule, and simplify the test/remove the comments re: deleting the wires after typechecking?

thiskappaisgrey commented:
So the original issue is that the expression will never get a type without the `Wire` typing rule. I just assumed that the `Wire`s need to not get deleted before we do the type checking but maybe my assumption is wrong. I'll give your suggestions a try.

thiskappaisgrey commented:
@gussmith23 - I removed the `Wire` rule because you don't need it. If the `Wire`s are meant to be deleted BEFORE applying the `typing` rewrites, then the rule would have no effect anyways. I also added the rule for `Shr` as well. Thanks.

gussmith23 commented:
Great!! but wait...the Wire rule is actually needed! Sure, Wires _should_ be deleted, but it's not required that they be deleted. So for now I think we should keep the rule (esp. because the missing rule could just cause missing types like you experienced, which is very frustrating!)

thiskappaisgrey commented:
Ok, I added the wire rule back.

gussmith23 commented:
Thanks @thiskappaisgrey for your first contribution!!! 🎉  Let me know what other questions you have or what other features you might need.
